plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts

```diff
@@ -49,23 +49,70 @@ import {
 import { TeamMember as AdoTeamMember } from 'azure-devops-node-api/interfaces/common/VSSInterfaces';
 import { Logger } from 'winston';
 import { PolicyEvaluationRecord } from 'azure-devops-node-api/interfaces/PolicyInterfaces';
-import { WebApi } from 'azure-devops-node-api';
+import {
+  WebApi,
+  getBearerHandler,
+  getPersonalAccessTokenHandler,
+} from 'azure-devops-node-api';
 import {
   TeamProjectReference,
   WebApiTeam,
 } from 'azure-devops-node-api/interfaces/CoreInterfaces';
 import { UrlReader } from '@backstage/backend-common';
+import { Config } from '@backstage/config';
+import {
+  DefaultAzureDevOpsCredentialsProvider,
+  ScmIntegrations,
+} from '@backstage/integration';
 
 /** @public */
 export class AzureDevOpsApi {
-  public constructor(
-    private readonly logger: Logger,
-    private readonly webApi: WebApi,
-    private readonly urlReader: UrlReader,
-  ) {}
+  private readonly logger: Logger;
+  private readonly urlReader: UrlReader;
+  private readonly config: Config;
+  private readonly credentialsProvider: DefaultAzureDevOpsCredentialsProvider;
+
+  private constructor(logger: Logger, urlReader: UrlReader, config: Config) {
+    this.logger = logger;
+    this.urlReader = urlReader;
+    this.config = config;
+
+    const scmIntegrations = ScmIntegrations.fromConfig(this.config);
+    this.credentialsProvider =
+      DefaultAzureDevOpsCredentialsProvider.fromIntegrations(scmIntegrations);
+  }
+
+  static fromConfig(
+    config: Config,
+    options: { logger: Logger; urlReader: UrlReader },
+  ) {
+    return new AzureDevOpsApi(options.logger, options.urlReader, config);
+  }
+
+  private async getWebApi(host?: string, org?: string): Promise<WebApi> {
+    const validHost = host ?? this.config.getString('azureDevOps.host');
+    const validOrg = org ?? this.config.getString('azureDevOps.organization');
+    const url = `https://${validHost}/${validOrg}`;
+
+    const credentials = await this.credentialsProvider.getCredentials({
+      url,
+    });
+
+    // TODO:(awanlin) use `getHandlerFromToken` once we no longer
+    // need to support the 'azureDevOps.token' fallback config value
+    const authHandler =
+      credentials?.type === 'pat'
+        ? getPersonalAccessTokenHandler(
+            credentials!.token ?? this.config.getString('azureDevOps.token'),
+          )
+        : getBearerHandler(credentials!.token);
```

Comment:
at this point `credentials` might be `undefined`. I'm not familiar with the logic behind credentials so I'm not sure whether we need some extra error handling here 🤔

Reply:
Yeah, I think when I revisited this from prior comments I didn't have a clear model of this in my head. The latest refactoring handles this much better IMO and ended up easier to read 👍